src/szk.rs

```diff
@@ -0,0 +1,574 @@
+            if let Some(leader_seed) = leader_seed_opt_copy {
+                let leader_joint_rand_part =
+                    self.derive_joint_rand_part(&leader_seed, leader_input_share, nonce)?;
+                let helper_joint_rand_part =
+                    self.derive_joint_rand_part(&helper_seed, helper_input_share, nonce)?;
+                let (_jrs, jr) =
+                    self.derive_joint_rand(&leader_joint_rand_part, &helper_joint_rand_part);
+                (
+                    Some((leader_seed, helper_joint_rand_part)),
+                    Some(leader_joint_rand_part),
+                    jr,
+                )
+            } else {
+                (None, None, Vec::<T::Field>::new())
+            };
+
+        let prove_rand = self.derive_prove_rand(&prove_rand_seed);
+        let mut leader_proof_share =
+            self.typ
+                .prove(encoded_measurement, &prove_rand, &joint_rand)?;
+
+        // Generate the proof shares.
+        for (x, y) in leader_proof_share
+            .iter_mut()
+            .zip(self.derive_helper_proof_share(&helper_seed))
+        {
+            *x -= y;
+        }
+
+        // Construct the output messages.
+        let leader_proof_share = SzkProofShare::Leader {
+            uncompressed_proof_share: leader_proof_share,
+            leader_blind_and_helper_joint_rand_part,
+        };
+        let helper_proof_share = SzkProofShare::Helper {
+            proof_share_seed_and_blind: helper_seed.clone(),
+            leader_joint_rand_part,
+        };
+        Ok([leader_proof_share, helper_proof_share])
+    }
+
+    fn query(
+        &self,
+        input_share: &[T::Field],
+        proof_share: SzkProofShare<T::Field, SEED_SIZE>,
+        verify_key: &[u8; SEED_SIZE],
+        nonce: &[u8; 16],
+    ) -> Result<SzkQueryShare<T::Field, SEED_SIZE>, SzkError> {
+        let query_rand = self.derive_query_rand(verify_key, nonce);
+
+        let is_leader = match proof_share {
+            SzkProofShare::Leader { .. } => true,
+            SzkProofShare::Helper { .. } => false,
+        };
+
+        let flp_proof_share = match proof_share {
+            SzkProofShare::Leader {
+                ref uncompressed_proof_share,
+                ..
+            } => Cow::Borrowed(uncompressed_proof_share),
+            SzkProofShare::Helper {
+                ref proof_share_seed_and_blind,
+                ..
+            } => Cow::Owned(self.derive_helper_proof_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, joint_rand_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, peer_joint_rand_part) = match proof_share {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    leader_blind_and_helper_joint_rand_part,
+                } => leader_blind_and_helper_joint_rand_part
+                    .expect("jr_len > 0 so there is joint_randomness"),
```

Comment:
"jr_len" is not a variable name. A more helpful panic message would be something like this: "leader_blind_and_helper_joint_rand_part should be set"

Likewise below.